File: ag-compiler/src/parser.rs
use std::mem::replace;

use crate::{
    repr::{CanParse, ModuleRepr},
    scanner::{ScanError, Scanner},
    token::Token,
};

pub struct Parser<'src, 'repr, R: CanParse> {
    source: &'src String,
    scanner: Scanner<'src>,
    repr: &'repr mut R,
    current: Token,
    prev: Token,
    is_main_file: bool,
}
impl<'src, 'repr, R: CanParse> Parser<'src, 'repr, R> {
    pub fn new(source: &'src String, repr: &'repr mut R, is_main_file: bool) -> Self {
        Self {
            source,
            scanner: Scanner::new(source),
            repr,
            current: Token::eof(0, 0..0),
            prev: Token::eof(0, 0..0),
            is_main_file,
        }
    }

    fn next(&mut self) -> Result<&Token, ScanError> {
        let new = self.scanner.scan()?;
        let _ = replace(&mut self.prev, replace(&mut self.current, new));
        Ok(&self.current)
    }
    pub fn parse(&mut self) -> bool {
        true
    }
}

File: AgateCS/Compiler/Parser.cs
using AgateCS.Scanner;
using AgateCS.Scanner.ScanUtils;
using AgateCS.Utils;

namespace AgateCS.Compiler
{
    public static class Parser
    {
        private static class Tokens
        {
            public static Token Current = Token.Eof();
            public static Token Previous = Token.Eof();

            private static IScanResult GetToken() => Scanner.Scanner.ScanToken();
            public static Token Advance()
            {
                for (; ; )
                {
                    IScanResult result = GetToken();
                    if (result is Token token)
                    {
                        Previous = Current;
                        Current = token;
                        return Current;
                    }
                    else
                    {
                        Console.Error.WriteLine("Scan Error: ", result.Message());
                    }
                }
            }
        }
        
        private static bool IsMain = false;
        private static string Source = "";
        private static ModuleAst Module = new("");
        private static bool HadError = false;

        public static void Init(string source, bool isMain = false)
        {
            Source = source;
            Scanner.Scanner.ReInit(source);
            IsMain = isMain;
        }
        private static Token Advance() => Tokens.Advance();

        public static void ParseItem()
        {
            throw new NotImplementedException();
        }
        public static bool Parse(ModuleAst module, bool isMain)
        {
            IsMain = isMain;
            HadError = false;
            Module = module;
            Advance();

            while (Tokens.Current.Type != TokenType.EOF)
            {
                ParseItem();
            }

            return HadError;
        }
    }
}
